### PR TITLE
[WIP DO NOT MERGE] Allow users to share templates with Accounts or Projects through the

### PR DESCRIPTION
**Pull request by rhtyd**
Problem: CloudStack users want the ability to share templates with other accounts or projects through the UI.

Root Cause: The ‘updateTemplatePermissions’ is only available via API and needs to be integrated with the UI in order to allow users to share templates with other accounts within the same domain, and also with projects that the sharing account has access to.

Solution: The CloudStack UI template view has been extended to include a button that allows users to update template permissions through the ‘updateTemplatePermissions’ API. This API supports 3 operations i.e. add, remove or reset. The reset operation does not require any parameters and it simply removes all permissions for a template. The add and remove parameters require either accounts or projects to be specified for sharing or stop sharing templates. The ‘listAccounts’ API has been extended to allow users to see all the accounts within the same domain if the global setting ‘allow.view.all.domain.accounts’ is set to true. This allows the accounts selection to be a multi-selection.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):
Reset Operation hides the Accounts and Project ID fields as this operation does not require any additional parameters to be set.

Operation = Reset; Account and Project fields not displayed
image

Operation = Add/Remove , global setting: 'allow.user.view.all.domain.accounts' = false;
image

Operation = Add , global setting: 'allow.user.view.all.domain.accounts' = true; Domain has 3 accounts in total (bt1, bt2 and bt3). Logged in user is bt1 and so it is not shown on the dropdown list.
image

Operation = Remove , global setting: 'allow.user.view.all.domain.accounts' = true; Domain has 3 accounts in total (bt1, bt2 and bt3). The current template was only shared with bt2 by bt1 so dropdown menu only shows accounts that the current template has been shared with.
image

## How Has This Been Tested?
- Create 2 Accounts (acc1 and acc2) within the same domain.
- Log in to the CloudStack UI with acc1 user account and upload a private template for this user.
- Navigate to the newly created UI and click on the "Update Template Permissions" button.
- A new window should pop up, select "Add" for this window and input acc2 in the Accounts input field and then click OK.
- With CloudMonkey/CloudStack UI, check for shared templates using acc2 user account. The template from acc1 user account should now be visible to acc2 under shared templates.

### CODE REVIEW

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by anuragaw** [comment]
@blueorangutan package

**Comment by anuragaw** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@rhtyd conflict issues with this.
For a non-public template in ROOT domain, Root admin cannot share template to a project in one of the subdomain?
Screenshot from 2019-05-29 17-07-30
Screenshot from 2019-05-29 17-15-29
Screenshot from 2019-05-29 17-15-45

**Comment by rhtyd** [comment]
@shwstppr @anuragaw  can you fix the merge conflicts, thanks.

**Comment by anuragaw** [comment]
Fixed merge conflicts and cleaned up some code left over. @rhtyd , @shwstppr

**Comment by anuragaw** [comment]
> @rhtyd conflict issues with this.
> For a non-public template in ROOT domain, Root admin cannot share template to a project in one of the subdomain?
> Screenshot from 2019-05-29 17-07-30
> Screenshot from 2019-05-29 17-15-29
> Screenshot from 2019-05-29 17-15-45

We may have to ask @rhtyd or @PaulAngus if this is the expected behaviour or not.

**Comment by rhtyd** [comment]
@anuragaw can you address this bug if you've time today?

**Comment by anuragaw** [comment]
Looking now @rhtyd

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by anuragaw** [comment]
Fixed bug and rebased against master @rhtyd @shwstppr @borisstoyanov 

@shwstppr - can you retest the PR please? Thanks!

**Comment by anuragaw** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 1]
```diff
@@ -1541,14 +1539,23 @@ public boolean updateTemplateOrIsoPermissions(BaseUpdateTemplateOrIsoPermissions
...
             final List<String> accountNamesFinal = accountNames;
             final List<Long> accountIds = new ArrayList<Long>();
             Transaction.execute(new TransactionCallbackNoReturn() {
                 @Override
                 public void doInTransactionWithoutResult(TransactionStatus status) {
                     for (String accountName : accountNamesFinal) {
-                        Account permittedAccount = _accountDao.findActiveAccount(accountName, domain.getId());
+                        Account permittedAccount = null;
+                        if (isCallerRootAdmin) {
+                            List<AccountVO> accountsWthName = _accountDao.findAccountsLike(accountName);
```
Comment (shwstppr) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java]: @anuragaw  nit: Is variable name `accountsWthName` without **i** intentional? Same in block for remove

Comment (anuragaw) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java]: changing

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 2]
```diff
@@ -1541,14 +1539,23 @@ public boolean updateTemplateOrIsoPermissions(BaseUpdateTemplateOrIsoPermissions
...
+                            if (accountsWthName.size() > 0) {
+                                permittedAccount = accountsWthName.get(0);
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java]: @anuragaw could this cause regression or issues, this is always picking up the first account in the list; should it be something like - if user is root admin, then all enable/permit for all accounts matching the name using something like `accountIds.add(permittedAccount.getId());` in the if block; and refactor suitably?

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 3]
```diff
@@ -1578,7 +1585,15 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
         } else if ("remove".equalsIgnoreCase(operation)) {
             List<Long> accountIds = new ArrayList<Long>();
             for (String accountName : accountNames) {
-                Account permittedAccount = _accountDao.findActiveAccount(accountName, domain.getId());
+                Account permittedAccount = null;
+                if (isCallerRootAdmin) {
+                    List<AccountVO> accountsWthName = _accountDao.findAccountsLike(accountName);
+                    if (accountsWthName.size() > 0) {
+                        permittedAccount = accountsWthName.get(0);
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java]: @anuragaw could this cause regression or issues, this is always picking up the first account in the list; should it be something like - if user is root admin, then all enable/permit for all accounts matching the name using something like `accountIds.add(permittedAccount.getId());` in the if block; and refactor suitably?

File: server/src/main/java/com/cloud/api/ApiResponseHelper.java [hunk 4]
```diff
@@ -1789,8 +1795,23 @@ public TemplatePermissionsResponse createTemplatePermissionsResponse(ResponseVie
...
         List<String> regularAccounts = new ArrayList<String>();
+        final boolean isRootAdmin = _accountMgr.isRootAdmin(CallContext.current().getCallingAccountId());
         for (String accountName : accountNames) {
-            Account account = ApiDBUtils.findAccountByNameDomain(accountName, templateOwner.getDomainId());
+            // Root Admins should be allowed to access permissions for all accounts across domains
+            Account account = null;
+            if (isRootAdmin) {
+                List<AccountVO> accounts = accountDao.findAccountsLike(accountName);
+                if (accounts != null && accounts.size() > 0) {
+                    account = accounts.get(0);
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/api/ApiResponseHelper.java]: @anuragaw could this cause regression or issues, this is always picking up the first account in the list; should it be something like - if user is root admin, then all enable/permit for all accounts matching the name using something like `accountIds.add(permittedAccount.getId());` in the if block; and refactor suitably?

Comment (rhtyd) [reply server/src/main/java/com/cloud/api/ApiResponseHelper.java]: Should this find account using name and domain?

**Comment by rhtyd** [comment]
@anuragaw can you fix the conflicts? Thanks.

**Comment by anuragaw** [comment]
Sure @rhtyd - there is some more work to be done and I'll rebase with new code.

**Comment by anuragaw** [comment]
@blueorangutan package

**Comment by anuragaw** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java (hunk 2 above)
Comment (anuragaw) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java]: @rhtyd - good call. I looked at the implementation in more depth on Friday and observed the following -
1. Update template permissions uses account names (NOT uuids) to update permissions.
2. This means that the above API call is meant for domain specific sharing and hence implies a cross domain sharing restriction.
3. Double checked and confirmed that sharing templates VIA UI should also have same restrictions as updateTemplatePermissions API.

Hence I've reverted some changes and updated UI code to reflect these restrictions.

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java (hunk 3 above)
Comment (anuragaw) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java]: No longer part of PR as above.

File: server/src/main/java/com/cloud/api/ApiResponseHelper.java (hunk 4 above)
Comment (anuragaw) [reply server/src/main/java/com/cloud/api/ApiResponseHelper.java]: No longer part of PR as above.

File: ui/css/cloudstack3.css [hunk 5]
```diff
@@ -12372,6 +12372,14 @@ div.ui-dialog div.autoscaler div.field-group div.form-container form div.form-it
   background-position: -35px -707px;
 }
 
+.shareTemplate .icon {
+  background-position: -165px -122px;
+}
+
+.shareTemplate .icon {
```
Comment (shwstppr) [inline ui/css/cloudstack3.css]: @anurag should this be` .shareTemplate:hover .icon {`?

**Comment by borisstoyanov** [comment]
@blueorangutan package

File: ui/css/cloudstack3.css (hunk 5 above)
Comment (anuragaw) [reply ui/css/cloudstack3.css]: Done.

**Comment by anuragaw** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@anuragaw @shwstppr @nvazquez please review

**Comment by anuragaw** [comment]
@rhtyd , @borisstoyanov , @PaulAngus - rebased off master.

**Comment by anuragaw** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Looks like Github has gone bonkers @borisstoyanov - I see no conflict via command line.

**Comment by rhtyd** [comment]
Conflicts should be fixed now cc @borisstoyanov

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by anuragaw** [comment]
@blueorangutan - package

**Comment by anuragaw** [comment]
@blueorangutan package

**Comment by anuragaw** [comment]
When admins share project owned templates we will now get a distinct message.

**Comment by anuragaw** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
LGTM with the known issue of #3481

**Pull request merged**